src/sid/update_states.py

```diff
@@ -202,34 +198,103 @@ def update_derived_state_variables(states, derived_state_variables):
+    reaching the maximum. Before the maximum is achieved the function is modeled as a
+    third-degree polynomial and afterwards as a linear function. The coefficients are
+    adjusted to the parameters and set in ``_get_waning_immunity_coefficients``.
+
+    Args:
+        params (pandas.DataFrame): See :ref:`params`.
+        days_since_event (pandas.Series): Series containing days since event occurred
+            for each individual in the state.
+        event (str): Reason for immunity. Must be in {"infection", "vaccination"}.
+
+    Returns:
+        immunity (pandas.Series): Adjusted immunity level.
+
+    """
+    coef = _get_waning_immunity_coefficients(params, event)
+    immunity = pd.Series(0, index=days_since_event.index, dtype=DTYPE_IMMUNITY)
+
+    before_maximum = (days_since_event > 0) & (
+        days_since_event < coef["time_to_reach_maximum"]
     )
-    immunity_from_vaccination = (
-        newly_immune_by_vaccine.astype(DTYPE_IMMUNITY)
-        * params.loc[("immunity", "immunity_level", "from_vaccination"), "value"]
+    after_maximum = days_since_event >= coef["time_to_reach_maximum"]
+
+    # increase immunity level for individuals who have not reached their maximum level
+    immunity[before_maximum] = coef["slope_before_maximum"] * (
+        days_since_event[before_maximum] ** 3
     )
 
-    levels = np.column_stack(
-        [immunity, immunity_from_infection, immunity_from_vaccination]
+    # decrease immunity level for individuals who are beyond their maximum level
+    immunity[after_maximum] = (
+        coef["intercept"]
+        + coef["slope_after_maximum"] * days_since_event[after_maximum]
     )
-    states["immunity"] = levels.max(axis=1)
-    return states
+
+    # make sure that immunity level is non-negative, which is not automatically enforced
+    # by the linear function specification
+    immunity = immunity.clip(lower=0)
+    return immunity
+
+
+def _get_waning_immunity_coefficients(
+    params: pd.DataFrame, event: str
+) -> Dict[str, float]:
+    """Transform high-level arguments for waning immunity to low-level coefficients.
+
+    Coefficients are calibrated to parameters in params.
+
+    Args:
+        params (pandas.DataFrame): See :ref:`params`.
+        event (str): Reason for immunity. Must be in {"infection", "vaccination"}.
+
+    Returns:
+        coef (Dict[str, float]): The coefficients.
+
+    """
+    maximum_immunity = params.loc[
+        ("immunity", "immunity_level", f"from_{event}"), "value"
+    ]
+    time_to_reach_maximum = params.loc[
+        ("immunity", "immunity_waning", f"time_to_reach_maximum_{event}"), "value"
+    ]
+    slope_after_maximum = params.loc[
+        ("immunity", "immunity_waning", f"slope_after_maximum_{event}"), "value"
+    ]
+
+    slope_before_maximum = maximum_immunity / (time_to_reach_maximum ** 3)
+    intercept = maximum_immunity - slope_after_maximum * time_to_reach_maximum
```

Comment:
This doesn't seem to be right to me. 

```python

maximum_immunity = 0.8
time_to_reach_maximum = 4

slope_before_maximum = maximum_immunity / (time_to_reach_maximum ** 3)
intercept = maximum_immunity - slope * time_to_reach_maximum

intercept + slope * 4 ** 3
```

Here I get 1.55, not 0.8 which is what we would want.

Reply:
The code is correct but not readable. The overall immunity curve ``f : [0, infinity] -> [0, 1]`` is defined via the cubic curve ``f_1(t) = slope_before_maximum * (t ** 3)`` for the case where ``t < time_to_reach_maximum`` and the linear part ``f_2(t) = intercept + slope_after_maximum * t`` for the case where ``t >= time_to_reach_maximum``.

I will rename ``intercept`` to ``intercept_after_maximum`` to make it more clear. In your number example you would simply leave out the intercept and then you get 0.8